### PR TITLE
Modify ingester's limiter to work with ingest store.

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This PR modifies ingester's `Limiter` code to work with ingest store (partitions ring) as well as ingester ring. ~Based on top of https://github.com/grafana/mimir/pull/7411.~ Rebased on top of `main`.

#### Checklist

- [x] Tests updated.
- [na] Documentation added.
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [na] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/ingester/limiter.go [hunk 1]
```diff
@@ -137,16 +172,73 @@ func (l *Limiter) convertGlobalToLocalLimit(userID string, globalLimit int, minL
...
+
+func (ps *partitionRingLimiterStrategy) convertGlobalToLocalLimit(userID string, globalLimit int) int {
+	if globalLimit == 0 {
+		return 0
+	}
+
+	userShardSize := ps.getShardSize(userID)
+
+	pr := ps.partitionRingWatcher.PartitionRing()
+	activePartitionsCount := pr.ActivePartitionsCount()
```
Comment (pracucci) [inline pkg/ingester/limiter.go:214]: I was wondering if `pr.ActivePartitionsCount()` + logic below is mathematically equivalent to `pr.ShuffleShard().ActivePartitionsCount()`. I think it's the same. The latter is probably more accurate, but potentially slower, so I'm happy to keep your version. Would be nice a test that asserts the two returns the same numbers.

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ingester/limiter.go:214]: This is done in https://github.com/grafana/dskit/pull/493.

Comment (pstibrany) [reply pkg/ingester/limiter.go:214]: https://github.com/grafana/dskit/pull/493 adds `PartitionRing.ShuffleShardSize` method with the same computation, and also adds checks into existing tests to verify that it returns correct value.

File: pkg/ingester/limiter.go [hunk 2]
```diff
@@ -89,36 +82,78 @@ func (l *Limiter) maxMetadataPerUser(userID string) int {
...
 	// If the limit is disabled
 	if localLimit == 0 {
 		localLimit = math.MaxInt32
 	}
 
-	return localLimit
+	return max(localLimit, minLocalLimit)
+}
+
+func (l *Limiter) getShardSize(userID string) int {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ingester/limiter.go:102]: This is still in use on `main`, but usage is removed in https://github.com/grafana/mimir/pull/7411. Given that I've rebased PR back onto `main`, I will keep the method for now.

File: pkg/ingester/limiter.go (hunk 1 above)
Comment (pstibrany) [reply pkg/ingester/limiter.go]: I've updated this code to use newly introduced method from dskit.

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -405,14 +405,8 @@ func New(cfg Config, limits *validation.Overrides, ingestersRing ring.ReadRing,
...
-	// Init the limter and instantiate the user states which depend on it
-	i.limiter = NewLimiter(
-		limits,
-		ingestersRing,
-		cfg.IngesterRing.ReplicationFactor,
-		cfg.IngesterRing.ZoneAwarenessEnabled,
-		cfg.IngesterRing.InstanceZone,
-	)
+	var limiterStrategy limiterRingStrategy
+	limiterStrategy = newIngesterRingLimiterStrategy(ingestersRing, cfg.IngesterRing.ReplicationFactor, cfg.IngesterRing.ZoneAwarenessEnabled, cfg.IngesterRing.InstanceZone, i.limits.IngestionTenantShardSize)
```
Comment (pracucci) [inline pkg/ingester/ingester.go:409]: [nit] To make the intention more clear, I would move `newIngesterRingLimiterStrategy` in a `else` branch of `if ingestCfg := cfg.IngestStorageConfig; ingestCfg.Enabled {` so that we don't initialise the "wrong" and then override it when ingest storage is enabled.

File: pkg/ingester/limiter_test.go [hunk 4]
```diff
@@ -295,13 +335,97 @@ func runLimiterMaxFunctionTest(
...
+			globalLimit:               1000,
+			partitionStates:           []ring.PartitionState{ring.PartitionActive, ring.PartitionInactive, ring.PartitionInactive, ring.PartitionInactive, ring.PartitionInactive},
+			tenantPartitionsShardSize: 0,
+			expectedValue:             1000, // 1000 / 1
+		},
+		"limit is enabled, using all active partitions, only 2 are active": {
+			globalLimit:               1000,
+			partitionStates:           []ring.PartitionState{ring.PartitionActive, ring.PartitionActive, ring.PartitionInactive, ring.PartitionInactive, ring.PartitionInactive},
+			tenantPartitionsShardSize: 0,
+			expectedValue:             500, // 1000 / 2
```
Comment (pracucci) [inline pkg/ingester/limiter_test.go:374]: [nit] I like showing the intention, but maybe you can just replace the expected value with the comment content.

```suggestion
			expectedValue:             1000 / 2,
```

_Same comment applies in other cases of this test._

File: pkg/ingester/limiter_test.go [hunk 5]
```diff
@@ -449,14 +662,66 @@ func TestLimiter_AssertMaxSeriesPerUser(t *testing.T) {
...
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			pr := buildPartitionRingFromPartitionStates(testData.partitionStates...)
+
+			limits, err := validation.NewOverrides(validation.Limits{MaxGlobalSeriesPerUser: testData.maxGlobalSeriesPerUser}, nil)
+			require.NoError(t, err)
+
+			strategy := newPartitionRingLimiterStrategy(&partitionRingHolder{pr: pr}, limits.IngestionTenantShardSize)
+			limiter := NewLimiter(limits, strategy)
+			actual := limiter.IsWithinMaxSeriesPerUser("test", testData.series, testData.minLocalLimit)
```
Comment (pracucci) [inline pkg/ingester/limiter_test.go:718]: The 3rd param in this PR is not min local limit but the shard size. I think the cases in this test need to be redone.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:1]: The partition ring works with partition shards. When used together with the owned series service, in `Ingester.createTSDB()` we initialise `ownedSeriesShardSize:    i.limits.IngestionTenantShardSize(userID)` but we actually have to initialise it based on the right shard size. However, I guess that's part of your work to add partitions ring support to owned series service (which to my understanding will happen in a follow up PR). If so, then I'm good doing this change as part of that PR.

**Review by pracucci — APPROVED** [review]
LGTM (modulo a couple of comments). Thanks!

File: pkg/ingester/limiter.go
Comment (pstibrany) [reply pkg/ingester/limiter.go:1]: That's correct. I have changes to owned-series on local WIP branch, and will follow-up with PR soon.

File: pkg/ingester/limiter_test.go (hunk 5 above)
Comment (pstibrany) [reply pkg/ingester/limiter_test.go]: Thank you, good catch. I've fixed this test and also added test cases to `TestLimiter_AssertMaxSeriesPerUser` to pass non-zero shard size.

File: pkg/ingester/ingester.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ingester/ingester.go]: Done in https://github.com/grafana/mimir/pull/7424/commits/9811e539683e5ab52d353d1e3d0533992561ed60.

File: pkg/ingester/limiter_test.go (hunk 4 above)
Comment (pstibrany) [reply pkg/ingester/limiter_test.go]: Done in https://github.com/grafana/mimir/pull/7424/commits/9811e539683e5ab52d353d1e3d0533992561ed60.

**Review by pracucci — APPROVED** [review]
Thanks!

File: pkg/ingester/limiter.go [hunk 6]
```diff
@@ -92,10 +82,15 @@ func (l *Limiter) maxMetadataPerUser(userID string) int {
 	return l.convertGlobalToLocalLimitOrUnlimited(userID, l.getShardSize(userID), l.limits.MaxGlobalMetricsWithMetadataPerUser)
 }
 
+func (l *Limiter) maxExemplarsPerUser(userID string) int {
+	// We don't use `convertGlobalToLocalLimitOrUnlimited`, because we don't want "unlimited" part. 0 means disabled.
+	return l.ringStrategy.convertGlobalToLocalLimit(l.getShardSize(userID), l.limits.MaxGlobalExemplarsPerUser(userID))
```
Comment (pracucci) [inline pkg/ingester/limiter.go:87]: If we're unable to convert the global limit to local limit (e.g. because the ingesters ring client hasn't updated yet and see no ingesters yet) then we return 0, which means exemplars are disabled.

**Comment by pstibrany** [comment]
Thanks for your investigation and fix for flaky integration tests @pracucci.

**Comment by pracucci** [comment]
Fixes https://github.com/grafana/mimir/issues/7448
